### PR TITLE
Add recipe class

### PR DESCRIPTION
**Pull request by SeenFang**

### CODE REVIEW

File: src/main/java/decodex/data/exception/RecipeException.java [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package decodex.data.exception;
+
+/**
+ * This exception is raised when an error during recipe execution occurs.
+ */
+public class RecipeException extends Exception {
+
+    public static final String EMPTY_RECIPE_MESSAGE = "[x] There is nothing to be removed.";
```
Comment (arraysius) [inline src/main/java/decodex/data/exception/RecipeException.java:8]: Perhaps the message can be "The recipe is currently empty. There is no module to be removed." to be more descriptive.

File: src/main/java/decodex/recipes/Recipe.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        logger.fine(String.format("[Recipe %s] Added module %s", name, module.getName()));
+    }
+
+    /**
+     * Removes the latest module that was added to the list.
+     *
+     * @return Module that was popped from the list
+     */
+    public Module pop() throws RecipeException {
+        if (moduleList.size() <= 0) {
```
Comment (arraysius) [inline src/main/java/decodex/recipes/Recipe.java:51]: Consider using `moduleList.isEmpty()` instead. It does not change the logic of the code but may help with readability.

**Review by arraysius — APPROVED** [review]
LGTM

File: src/main/java/decodex/recipes/Recipe.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    public void push(Module module) {
+        moduleList.add(module);
+        logger.fine(String.format("[Recipe %s] Added module %s", name, module.getName()));
+    }
+
+    /**
+     * Removes the latest module that was added to the list.
+     *
+     * @return Module that was popped from the list
+     */
```
Comment (rizemon) [inline src/main/java/decodex/recipes/Recipe.java:50]: Missing full-stop at the end of this line

File: src/test/java/decodex/recipes/RecipeTest.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+    @Test
+    void pop_singleModule_latestModule() throws RecipeException {
+        Recipe recipe = new Recipe("testRecipe");
+        Module module = new HexDecoder();
+        recipe.push(module);
+
+        Module poppedModule = recipe.pop();
+
+        assertEquals(poppedModule, poppedModule);
```
Comment (rizemon) [inline src/test/java/decodex/recipes/RecipeTest.java:34]: I think perhaps you wanted to compare `module` and `poppedModule`?

File: src/main/java/decodex/recipes/Recipe.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+public class Recipe {
+
+    private Logger logger = Decodex.logger;
+
+    private String name;
+    private ArrayList<Module> moduleList = new ArrayList<>();
+
+    public Recipe(String name) {
+        this.name = name;
+    }
```
Comment (rizemon) [inline src/main/java/decodex/recipes/Recipe.java:22]: Perhaps write `Instantiates a new recipe of modules`?

File: src/main/java/decodex/recipes/Recipe.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+/**
+ * The Recipe class manages a list of Modules to be run sequentially.
+ */
+public class Recipe {
+
+    private Logger logger = Decodex.logger;
+
+    private String name;
+    private ArrayList<Module> moduleList = new ArrayList<>();
```
Comment (rizemon) [inline src/main/java/decodex/recipes/Recipe.java:18]: Perhaps move the instantiation of the `moduleList` into the constructor?

File: src/main/java/decodex/recipes/Recipe.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+        this.name = name;
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    public void setName(String name) {
+        this.name = name;
+    }
```
Comment (rizemon) [inline src/main/java/decodex/recipes/Recipe.java:30]: Perhaps check if the recipe name is blank and throw an exception? Also if you are adding the check, make sure to create a test case for this too

File: src/main/java/decodex/recipes/Recipe.java [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+    public void push(Module module) {
+        moduleList.add(module);
+        logger.fine(String.format("[Recipe %s] Added module %s", name, module.getName()));
+    }
+
+    /**
+     * Removes the latest module that was added to the list.
+     *
+     * @return The module that was popped from the list.
+     */
```
Comment (rizemon) [inline src/main/java/decodex/recipes/Recipe.java:57]: Perhaps add a Javadoc for the exception?

**Review by rizemon — APPROVED** [review]
LGTM

**Pull request merged**